rpc_util.go

```diff
@@ -307,40 +309,45 @@ func encode(c Codec, msg interface{}, cp Compressor, cbuf *bytes.Buffer, outPayl
 			outPayload.Data = b
 			outPayload.Length = len(b)
 		}
+
 		if cp != nil {
+			// pre-alloc place
+			cbuf.Write(make([]byte, 5))
 			if err := cp.Do(cbuf, b); err != nil {
 				return nil, Errorf(codes.Internal, "grpc: error while compressing: %v", err.Error())
 			}
 			b = cbuf.Bytes()
+			length = uint(len(b) - 5)
+		} else {
+			length = uint(len(b))
 		}
-		length = uint(len(b))
 	}
+
 	if length > math.MaxUint32 {
 		return nil, Errorf(codes.ResourceExhausted, "grpc: message too large (%d bytes)", length)
 	}
 
-	const (
-		payloadLen = 1
-		sizeLen    = 4
-	)
-
-	var buf = make([]byte, payloadLen+sizeLen+len(b))
+	if cp != nil {
+		// Write payload format
+		b[0] = byte(compressionMade)
+		// Write length of b into buf
+		binary.BigEndian.PutUint32(b[1:], uint32(length))
+		if outPayload != nil {
+			outPayload.WireLength = len(b)
+		}
+		return b, nil
+	}
 
+	var buf []byte
```

Comment:
You are right. I thought buf = b will assign new alloc for buf, but it turns out it doesn't.